crates/cow-amm/src/registry.rs

```diff
@@ -0,0 +1,138 @@
+use {
+    crate::{CowAmm, Deployment},
+    ethrpc::current_block::{BlockRetrieving, CurrentBlockStream, RangeInclusive},
+    shared::{
+        event_handling::{EventHandler, EventRetrieving, EventStoring},
+        maintenance::{Maintaining, ServiceMaintenance},
+    },
+    std::{
+        any::TypeId,
+        collections::{BTreeMap, HashMap},
+        sync::Arc,
+    },
+    tokio::sync::{Mutex, RwLock},
+};
+
+/// CoW AMM indexer which stores events in-memory.
+#[derive(Clone)]
+pub struct Registry {
+    cow_amms: Arc<RwLock<BTreeMap<u64, Arc<dyn CowAmm>>>>,
```

Comment:
The clone is 100 % needed because you need to clone `self` for every new listener (because `self` is the one implementing the indexing and it is needed by the `EventHandler`

Reply:
And about having an inner struct of keeping the fields with `Arc<>`. I personally think it is a matter of style/preference. Have the inner struct is lighter in terms of memory (it needs just one `Arc<>` to conquer them all), but then it needs to do all the time `self.0`, and the struct content is not appearing in the struct definition (harder to read).